src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -37,6 +46,270 @@
 
 static char * PartitionBound(Oid partitionId);
 static Relation try_relation_open_nolock(Oid relationId);
+static List * CreateFixPartitionConstraintsTaskList(Oid relationId);
+static List * WorkerFixPartitionConstraintCommandList(Oid relationId, uint64 shardId,
+													  List *checkConstraintList);
+static List * CheckConstraintNameListForRelation(Oid relationId);
+static bool RelationHasConstraint(Oid relationId, char *constraintName);
+static char * RenameConstraintCommand(Oid relationId, char *constraintName,
+									  char *newConstraintName);
+
+
+PG_FUNCTION_INFO_V1(fix_pre_citus10_partitioned_table_constraint_names);
+PG_FUNCTION_INFO_V1(worker_fix_pre_citus10_partitioned_table_constraint_names);
+
+
+/*
+ * fix_pre_citus10_partitioned_table_constraint_names fixes the constraint names of
+ * partitioned table shards on workers.
+ *
+ * Constraint names for partitioned table shards should have shardId suffixes if and only
+ * if they are unique or foreign key constraints. We mistakenly appended shardIds to
+ * constraint names on ALTER TABLE dist_part_table ADD CONSTRAINT .. queries prior to
+ * Citus 10. fix_pre_citus10_partitioned_table_constraint_names determines if this is the
+ * case, and renames constraints back to their original names on shards.
+ */
+Datum
+fix_pre_citus10_partitioned_table_constraint_names(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	EnsureCoordinator();
+
+	if (!PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errmsg("could not fix partition constraints: "
+							   "relation does not exist or is not partitioned")));
+	}
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("fix_pre_citus10_partitioned_table_constraint_names can "
+							   "only be called for distributed partitioned tables")));
+	}
+
+	List *taskList = CreateFixPartitionConstraintsTaskList(relationId);
+	bool localExecutionSupported = true;
+	ExecuteUtilityTaskList(taskList, localExecutionSupported);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * worker_fix_pre_citus10_partitioned_table_constraint_names fixes the constraint names on a worker given a shell
+ * table name and shard id.
+ */
+Datum
+worker_fix_pre_citus10_partitioned_table_constraint_names(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	int64 shardId = PG_GETARG_INT32(1);
+	text *constraintNameText = PG_GETARG_TEXT_P(2);
+
+	if (!PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errmsg("could not fix partition constraints: "
+							   "relation does not exist or is not partitioned")));
+	}
+
+	char *constraintName = text_to_cstring(constraintNameText);
+	char *shardIdAppendedConstraintName = pstrdup(constraintName);
```

Comment:
Minor: `shardConstraintName` could be better than the current name. Currently we append shard ids to names but it can change in the future, ideally the method should also be renamed `AppendShardIdToName` -> `ConstructShardId`. That shouldn't be done in this PR but we can safely omit the `IdAppended` part from the current name.

Reply:
The purpose of this function is to remove incorrectly placed shardId suffixes in earlier versions of Citus. That's why I think we should keep the name as is. Even if we change how we construct shard names in future releases, the purpose of this function is to fix naming issues that happened before Citus 10

Reply:
You are right, good point :)